client/allocrunner/networking_cni.go

```diff
@@ -123,8 +158,202 @@ func (c *cniNetworkConfigurator) Setup(ctx context.Context, alloc *structs.Alloc
 		c.logger.Debug("received result from CNI", "result", string(resultJSON))
 	}
 
-	return c.cniToAllocNet(res)
+	allocNet, err := c.cniToAllocNet(res)
+	if err != nil {
+		return nil, err
+	}
 
+	// prepend the Consul DNS to the nameservers, if we have it; we don't need
+	// the port because the iptables rule redirects port 53 traffic to it
+	if tproxyArgs != nil && tproxyArgs.ConsulDNSIP != "" {
+		if allocNet.DNS == nil {
+			allocNet.DNS = &structs.DNSConfig{
+				Servers:  []string{},
+				Searches: []string{},
+				Options:  []string{},
+			}
+		}
+		allocNet.DNS.Servers = append([]string{tproxyArgs.ConsulDNSIP},
+			allocNet.DNS.Servers...)
+	}
```

Comment:
(Q in slack about combining nameservers like this)

Reply:
Summary of that discussion: we should overwrite the nameserver (unless `NoDNS` is set, which isn't the case in this code path), because the port 53 redirect rule makes it impossible to use alternate nameservers. We'll validate that `network.dns.namesevers` aren't set on job submit in a separate PR.